{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a23ebc2-f4c6-48b5-8af6-758ccebe78b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import data from FASTQ pass folder on MinION device. Only files of used barcodes were retained, then sorted into folders by primer\n",
    "import os\n",
    "import glob\n",
    "\n",
    "WB_S_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass/18S\"\n",
    "WB_R_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass/rbcL\"\n",
    "FC_S_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass/18S\"\n",
    "FC_R_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass/rbcL\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "908ab220-2a81-410f-a0fb-4b4e32c2ceee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creates new concatenated files (one for each barcode) of all files in a barcode folder\n",
    "os.chdir(WB_S_locations)\n",
    "\n",
    "for file in WB_S_list:\n",
    "    wildcard = file[:-12] + '*' + '.fastq'\n",
    "    final_name = file[:-13] + '.fastq'\n",
    "    !cat $wildcard > $final_name\n",
    "\n",
    "os.chdir(WB_R_locations)\n",
    "\n",
    "for file in WB_R_list:\n",
    "    wildcard = file[:-12] + '*' + '.fastq'\n",
    "    final_name = file[:-13] + '.fastq'\n",
    "    !cat $wildcard > $final_name\n",
    "    \n",
    "os.chdir(FC_S_locations)\n",
    "\n",
    "for file in FC_S_list:\n",
    "    wildcard = file[:-12] + '*' + '.fastq'\n",
    "    final_name = file[:-13] + '.fastq'\n",
    "    !cat $wildcard > $final_name\n",
    "    \n",
    "os.chdir(FC_R_locations)\n",
    "\n",
    "for file in FC_R_list:\n",
    "    wildcard = file[:-12] + '*' + '.fastq'\n",
    "    final_name = file[:-13] + '.fastq'\n",
    "    !cat $wildcard > $final_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e4ec5b7-4c87-49ce-a319-7425353ceb1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#removes old, unconcatenated files\n",
    "old_WB_S_files = glob.glob(os.path.join(WB_S_locations, '*_filt.fastq'))\n",
    "old_WB_R_files = glob.glob(os.path.join(WB_R_locations, '*_filt.fastq'))\n",
    "old_FC_S_files = glob.glob(os.path.join(FC_S_locations, '*_filt.fastq'))\n",
    "old_FC_R_files = glob.glob(os.path.join(FC_R_locations, '*_filt.fastq'))\n",
    "for file in old_WB_S_files:\n",
    "    os.remove(file)\n",
    "for file in old_WB_R_files:\n",
    "    os.remove(file)\n",
    "for file in old_FC_S_files:\n",
    "    os.remove(file)\n",
    "for file in old_FC_R_files:\n",
    "    os.remove(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624bbb79-c0b6-49b1-a46d-eeffe3615a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reassignes locations, creates new lists, and confirms correct file numbers\n",
    "WB_S_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass/18S\"\n",
    "WB_R_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass/rbcL\"\n",
    "\n",
    "WB_S_list = glob.glob(os.path.join(WB_S_locations, '*.fastq')) \n",
    "WB_R_list = glob.glob(os.path.join(WB_R_locations, '*.fastq')) \n",
    "print('Total Number of WB 18S Files: ',len(WB_S_list))\n",
    "print('Total Number of WB rbcL Files: ',len(WB_R_list))\n",
    "\n",
    "FC_S_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass/18S\"\n",
    "FC_R_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass/rbcL\"\n",
    "\n",
    "FC_S_list = glob.glob(os.path.join(FC_S_locations, '*.fastq')) \n",
    "FC_R_list = glob.glob(os.path.join(FC_R_locations, '*.fastq')) \n",
    "print('Total Number of FC 18S Files: ',len(FC_S_list))\n",
    "print('Total Number of FC rbcL Files: ',len(FC_R_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8cb705-14b4-46a2-9e03-ea47939cec28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#testing file naming convention\n",
    "os.chdir(WB_R_locations)\n",
    "for file in WB_R_list:\n",
    "    new_file_name = file[:-13] + '.fastq'\n",
    "    print(file)\n",
    "    print(new_file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df96b02b-04eb-454b-82b3-46a3f8e5e6e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "##runs cutadapt targeting forward and reverse primers, for 18S and rbcL, for data from both flow cell runs\n",
    "\n",
    "#runs cutadapt on 18S WB reads for forward adapter (for 18S-528F)\n",
    "os.chdir(WB_S_locations) # go to the directory of where files so when run command line saves them there \n",
    "for file in WB_S_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_18S_F.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g GCGGTAATTCCAGCTCCAA...GTAGGTGAACCTGCAGAAGGATCA --discard-untrimmed -o $new_file_name $file\n",
    "\n",
    "#runs cutadapt on 18S WB reads for reverse adapter (for 18S-1528R)\n",
    "for file in WB_S_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_18S_R.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g TGATCCTTCTGCAGGTTCACCTAC...TTGGAGCTGGAATTACCGC --discard-untrimmed -o $new_file_name $file\n",
    "\n",
    "#runs cutadapt on 18S FC reads for forward adapter (for 18S-528F)\n",
    "os.chdir(FC_S_locations) # go to the directory of where files so when run command line saves them there \n",
    "for file in FC_S_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_18S_F.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g GCGGTAATTCCAGCTCCAA...GTAGGTGAACCTGCAGAAGGATCA --discard-untrimmed -o $new_file_name $file\n",
    "    \n",
    "#runs cutadapt on 18S FC reads for reverse adapter (for 18S-1528R)\n",
    "for file in FC_S_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_18S_R.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g TGATCCTTCTGCAGGTTCACCTAC...TTGGAGCTGGAATTACCGC --discard-untrimmed -o $new_file_name $file\n",
    "    \n",
    "os.chdir(WB_R_locations)\n",
    "#runs cutadapt on rbcL WB reads for forward adapter\n",
    "for file in WB_R_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_rbcL_F.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g AAGGAGAAATHAATGTCT...GAGACTTACACAAGGTTGYTT --discard-untrimmed -o $new_file_name $file\n",
    "\n",
    "#runs cutadapt on rbcL WB reads for reverse adapter\n",
    "for file in WB_R_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_rbcL_R.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g AARCAACCTTGTGTAAGTCTC...AGACATTDATTTCTCCTT --discard-untrimmed -o $new_file_name $file\n",
    "\n",
    "os.chdir(FC_R_locations)\n",
    "#runs cutadapt on rbcL FC reads for forward adapter\n",
    "for file in FC_R_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_rbcL_F.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g AAGGAGAAATHAATGTCT...GAGACTTACACAAGGTTGYTT --discard-untrimmed -o $new_file_name $file\n",
    "\n",
    "#runs cutadapt on rbcL FC reads for reverse adapter\n",
    "for file in FC_R_list: # run through each file in the filelist\n",
    "    new_file_name = file[:-13] + '_rbcL_R.fastq' # includes pathway\n",
    "    # run command line prompt using file name and new file name\n",
    "    !cutadapt -g AARCAACCTTGTGTAAGTCTC...AGACATTDATTTCTCCTT --discard-untrimmed -o $new_file_name $file\n",
    "    \n",
    "    \n",
    "##old files were deleted manually in Finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a7543c4-91fe-4d66-925e-f68f40f9bda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Creates lists of filenames usable to run in NanoPlot. \n",
    "\n",
    "reads = os.listdir(WB_S_locations)\n",
    "print(*reads, sep = ' ')\n",
    "\n",
    "reads = os.listdir(WB_R_locations)\n",
    "print(*reads, sep = ' ')\n",
    "\n",
    "reads = os.listdir(WB_R_locations)\n",
    "print(*reads, sep = ' ')\n",
    "\n",
    "reads = os.listdir(WB_R_locations)\n",
    "print(*reads, sep = ' ')\n",
    "\n",
    "#I was unable to run these commands themselves in python, so they were run in a bash shell instead.\n",
    "#See NanoPlot file in Github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449bf11e-48ee-42c1-97b7-53ac9d453ba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Runs NanoFilt for all files using length parameters determined by nanofilt output\n",
    "#resets location names and lists\n",
    "WB_S_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass_copy/all_filtered_files/18S\"\n",
    "WB_R_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass_copy/all_filtered_files/rbcL\"\n",
    "\n",
    "WB_S_list = glob.glob(os.path.join(WB_S_locations, '*.fastq')) \n",
    "WB_R_list = glob.glob(os.path.join(WB_R_locations, '*.fastq')) \n",
    "\n",
    "FC_S_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass_copy/all_filtered_files/18S\"\n",
    "FC_R_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass_copy/all_filtered_files/rbcL\"\n",
    "\n",
    "FC_S_list = glob.glob(os.path.join(FC_S_locations, '*.fastq')) \n",
    "FC_R_list = glob.glob(os.path.join(FC_R_locations, '*.fastq')) \n",
    "\n",
    "os.chdir(WB_R_locations)\n",
    "for file in WB_R_list:\n",
    "    reads = file\n",
    "    new_file_name = file[:-6] + '_filt.fastq'\n",
    "    !NanoFilt -l 1451 --maxlength 1542 -q 7 $reads > $new_file_name\n",
    "\n",
    "os.chdir(WB_S_locations)\n",
    "for file in WB_S_list:\n",
    "    reads = file\n",
    "    new_file_name = file[:-6] + '_filt.fastq'\n",
    "    !NanoFilt -l 910 --maxlength 1350 -q 7 $reads > $new_file_name\n",
    "    \n",
    "os.chdir(FC_R_locations)\n",
    "for file in FC_R_list:\n",
    "    reads = file\n",
    "    new_file_name = file[:-6] + '_filt.fastq'\n",
    "    !NanoFilt -l 1444 --maxlength 1537 -q 7 $reads > $new_file_name\n",
    "    \n",
    "os.chdir(FC_S_locations)\n",
    "for file in FC_S_list:\n",
    "    reads = file\n",
    "    new_file_name = file[:-6] + '_filt.fastq'\n",
    "    !NanoFilt -l 910 --maxlength 1343 -q 7 $reads > $new_file_name\n",
    "\n",
    "##old files and files of failed reads were deleted manually in finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4dd9de6-161a-4cf7-8143-f2925aeb5a61",
   "metadata": {},
   "outputs": [],
   "source": [
    "###lists file names to create manifest csv files in excel to upload to Qiime2. \n",
    "WB_S_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass_copy/all_filtered_files/18S\"\n",
    "WB_R_locations = r\"/Users/mcanally23/Documents/WB_fastq_pass_copy/all_filtered_files/rbcL\"\n",
    "\n",
    "WB_S_list = glob.glob(os.path.join(WB_S_locations, '*.fastq'))\n",
    "for file in WB_S_list:\n",
    "    print(file)\n",
    "WB_R_list = glob.glob(os.path.join(WB_R_locations, '*.fastq'))\n",
    "for file in WB_R_list:\n",
    "    print(file)\n",
    "\n",
    "FC_S_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass_copy/all_filtered_files/18S\"\n",
    "FC_R_locations = r\"/Users/mcanally23/Documents/FC_fastq_pass_copy/all_filtered_files/rbcL\"\n",
    "\n",
    "FC_S_list = glob.glob(os.path.join(FC_S_locations, '*.fastq'))\n",
    "for file in FC_S_list:\n",
    "    print(file)\n",
    "FC_R_list = glob.glob(os.path.join(FC_R_locations, '*.fastq'))\n",
    "for file in FC_R_list:\n",
    "    print(file)\n",
    "\n",
    "#generated manifest files are included in the repository"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
